keanu-python/keanu/algorithm/sampling.py

```diff
@@ -17,31 +16,90 @@
 java_import(k.jvm_view(), "io.improbable.keanu.algorithms.mcmc.nuts.NUTS")
 java_import(k.jvm_view(), "io.improbable.keanu.algorithms.mcmc.Hamiltonian")
 
-algorithms = {
-    'metropolis': k.jvm_view().MetropolisHastings,
-    'NUTS': k.jvm_view().NUTS,
-    'hamiltonian': k.jvm_view().Hamiltonian
-}
+
+class PosteriorSamplingAlgorithm:
+
+    def __init__(self, sampler: JavaObject):
+        self._sampler = sampler
+
+    def get_sampler(self) -> JavaObject:
+        return self._sampler
+
+
+class MetropolisHastingsSampler(PosteriorSamplingAlgorithm):
+
+    def __init__(self,
+                 proposal_distribution: str = None,
+                 proposal_distribution_sigma: numpy_types = None,
+                 proposal_listeners=[],
+                 use_cache_on_rejection: bool = None):
+
+        if (proposal_distribution is None and len(proposal_listeners) > 0):
+            raise TypeError("If you pass in proposal_listeners you must also specify proposal_distribution")
+
+        builder: JavaObject = k.jvm_view().MetropolisHastings.builder()
+
+        if proposal_distribution is not None:
+            proposal_distribution_object = ProposalDistribution(
+                type_=proposal_distribution, sigma=proposal_distribution_sigma, listeners=proposal_listeners)
+            builder = builder.proposalDistribution(proposal_distribution_object.unwrap())
+
+        if use_cache_on_rejection is not None:
+            builder.useCacheOnRejection(use_cache_on_rejection)
+
+        super().__init__(builder.build())
+
+
+class HamiltonianSampler(PosteriorSamplingAlgorithm):
```

Comment:
We can just remove Hamilitonian. Might want to do it in another PR if it causes much extra effort.

Reply:
Yeah - I think it's best as its own PR.  Thanks for the review.